server/src/main/java/org/elasticsearch/search/aggregations/support/AggregatorSupplier.java

```diff
@@ -0,0 +1,22 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.search.aggregations.support;
+
+public interface AggregatorSupplier {
```

Comment:
I don't love having a no-op interface here, but it gives us a bit of type safety over just passing in any old object.  The problem here is that each aggregation family has a unique signature for how it constructs aggregators, so we can't have a common interface that means something.  This gives us a type that all those aggregation specific suppliers can inherit from, and we can be reasonably sure that it's a safe cast, since we trust that people aren't cross-registering aggregators.  And, if there were a cross-registered aggregator, it wouldn't work anyway, so failing if we detect a bad cast is ok too.

That said, this feels really clunky and I'm open to better options.  I spent a lot of time thinking about this problem, and this was the best solution I came up with.